File: LeetCode/DP/Golang/Leetcode764-largest-plus-sign.go
package sbgo

func orderOfLargestPlusSign(n int, mines [][]int) int {
  f,g := make([][]int, n), make([][]bool, n)
  for i := 0;i < n;i++ {
    f[i] = make([]int, n)
    g[i] = make([]bool, n)
    for j := 0;j < n;j++ {
      g[i][j] = true
    }
  }
  for _, x := range mines {
    g[x[0]][x[1]] = false
  }

  for i := 0;i < n;i++ {
    for j, s := 0, 0;j < n;j++ {
      if g[i][j] {
        s++
      } else {
        s = 0
      }
      f[i][j] = s
    }
  }
  min := func(i, j int) int {
    if i < j {
      return i
    }
    return j
  }
  for i := 0;i < n;i++ {
    for j, s := n - 1, 0;j >= 0;j-- {
      if g[i][j] {
        s++;
      } else {
        s = 0
      }
      f[i][j] = min(f[i][j], s)
    }
  }
  
  for j := 0;j < n;j++ {
    for i, s := 0, 0;i < n;i++ {
      if g[i][j] {
        s++;
      } else {
        s = 0
      }
      f[i][j] = min(f[i][j], s)
    }
  }
  
  for j := 0;j < n;j++ {
    for i, s := n - 1, 0;i >= 0;i-- {
      if g[i][j] {
        s++
      } else {
        s = 0
      }
      f[i][j] = min(f[i][j], s)
    }
  }
  res := 0
  max := func(i, j int) int {
    if i < j {
      return j
    }
    return i
  }
  for i := 0;i < n;i++ {
    for j := 0;j < n;j++ {
      res = max(f[i][j], res)
    }
  }
  return res
}

File: LeetCode/DP/Python/Leetcode764-largest-plus-sign.py
class Solution:
    def orderOfLargestPlusSign(self, n: int, mines: List[List[int]]) -> int:
        g, f = [[1] * n for _ in range(n)], [[0] * n for _ in range(n)]
        for x in mines: g[x[0]][x[1]] = 0
        for i in range(n):
            s = 0
            for j in range(n):
                if g[i][j]: s += 1
                else: s = 0
                f[i][j] = s

        for i in range(n):
            s = 0
            for j in range(n - 1, -1, -1):
                if g[i][j]: s += 1
                else: s = 0
                f[i][j] = min(f[i][j], s)

        for j in range(n):
            s = 0
            for i in range(n):
                if g[i][j]: s += 1
                else: s = 0
                f[i][j] = min(f[i][j], s)

        for j in range(n):
            s = 0
            for i in range(n - 1, -1, -1):
                if g[i][j]: s += 1
                else: s = 0
                f[i][j] = min(f[i][j], s)
        res = 0
        for i in range(n):
            for j in range(n):
                res = max(f[i][j], res)

        return res

